### PR TITLE
Add support for more advertisement fields. Handling ParameterMissing exceptions for ads.

### PR DESCRIPTION
**Pull request by jcass77**
Adds the 'imageUrl' and 'clickThroughUrl' fields for ad playlist items. With these two fields included we are pretty much extracting all of the ad metadata information that Pandora makes available.

I propose that we consider removing the optional ```includeBannerAd``` parameter. It contains an HTML fragment that most players will not be able to display, and retrieves and registers an additional ad token for a resource that is never actually utilised or made available to the end user.

### CODE REVIEW

**Comment by jcass77** [comment]
I've also added a check that ```station_id``` is always specified when retrieving the ad metadata - we'll need this later on to register the add so cannot be omitted.

Also added a few test cases for the same.

**Comment by jcass77** [comment]
Final change: it is probably better that ```get_playlist``` always returns ```Playlist``` so that the return type is not dependent on the results returned. 

The way the code was written it could in theory return the empty list [] if no results were found.

File: pandora/client.py [hunk 1]
```diff
@@ -263,6 +262,10 @@ def get_ad_item(self, station_id, ad_token):
         from .models.pandora import AdItem
 
         ad_metadata = self.get_ad_metadata(ad_token)
+        if not station_id:
```
Comment (mcrute) [inline pandora/client.py]: Can you validate this before the call to `get_ad_metadata`? There's no point in making an expensive network call if we don't have enough parameters to handle the response.

**Comment by mcrute** [comment]
Pending the one comment this looks ready to merge.

**Comment by jcass77** [comment]
Good point, done.

**Comment by jcass77** [comment]
The latest commit also proposes a fix for https://github.com/mcrute/pydora/issues/39. As the Pandora API does not contain this level of detail, I am not sure whether the fact that some ads do not have any tracking tokens assigned to them is 'normal', or if this is happening due to network issues on the client side.

I considered handling the exception in the client API or the AdItem model directly, but it is probably better if each consumer decides what it wants to do if ads cannot be registered?

Since most consumers should rely on ```iterate_forever``` to keep feeding them with tracks to play, it makes sense to have some basic error handling provided out of the box. The propose implementation will just return the ad for playback without insisting on registering it successfully.

**Comment by mcrute** [comment]
This and the other ad changes are released in [1.6.4](https://pypi.python.org/pypi/pydora/1.6.4)

**Comment by jcass77** [comment]
Thanks @mcrute. The ```iterate_forever``` function is starting to look a little messy (thanks to me). And I am still getting random ```ParameterMissing``` and ```StopIteration``` exceptions that I cannot explain.

I think Pandora doesn't bother to ensure that ad items always have tracking tokens, images, or audio urls available. So the assumption that ```AdItem``` will always be fully formed may be wrong.

I'm still debugging to try and figure it all out and will let you know if I uncover anything. Thanks for the merge.

**Comment by mcrute** [comment]
It might be better to just fire and forget with the `AdItem.register_ad` call, at least where we call in in `prepare_playback`, just catch and throw away all exceptions. There may be no ad tracking for certain kinds of ads, though I would find that surprising. I've been doing some testing here (Seattle) and have not found any ads that exhibit this behaviour nor have I run into any of these errors. Maybe we should add some detailed logging include payload logging in the transport and you can run with that for a while to see if we can track down the exact combination of things that cause this.

**Comment by jcass77** [comment]
I actually added some payload logging on the local copy that I use (just outputting everything in ```APITransport.__call__```) and it has helped a lot, so agree just leaving log statements to the debug logger might be good.

I think the other PR implements fire and forget but just for ```ParameterMissing```. We can probably expand on this incrementally as our understanding of the Pandora API and how it behaves develops.

But I would do it slowly to ensure we don't make any assumptions that turn out to be wrong. I've encountered a few internal server errors for example (don't know why yet), after which the server seemed to become unresponsive to further requests.

**Pull request merged**